unreal_pak/src/pakfile.rs

```diff
@@ -0,0 +1,154 @@
+//! PakFile data structure for reading large pak files
+
+use std::collections::HashMap;
+use std::io::{BufReader, BufWriter, Read, Seek, Write};
+
+use crate::error::PakError;
+use crate::file::{read_file, write_file};
+use crate::header::Header;
+use crate::index::{read_index, write_index, Footer, Index};
+use crate::pakversion::PakVersion;
+use crate::CompressionMethod;
+
+/// An Unreal pak file with it's data kept on disk and only read on demand.
+/// Good for working with very large files, but it has restrictions when it
+/// comes to writing files. For a more flexible alternative see <insert doc link> PakMemory
+#[derive(Debug)]
+pub struct PakFile<'data, R, W>
+where
+    &'data R: Read + Seek,
+    &'data W: Write + Seek,
+{
+    /// version of the pak file format this one is using
+    pub pak_version: PakVersion,
+    /// mount point (Unreal stuff)
+    pub mount_point: String,
+    /// the compression method preferred for this file
+    pub compression: CompressionMethod,
+    /// the compression block size
+    pub block_size: u32,
+    files: HashMap<String, Header>,
+    reader: Option<BufReader<&'data R>>,
+    writer: Option<BufWriter<&'data W>>,
+}
+
+impl<'data, R, W> PakFile<'data, R, W>
+where
+    &'data R: Read + Seek,
+    &'data W: Write + Seek,
+{
+    /// Creates a new `PakFile` configured to read files.
+    pub fn reader(reader: &'data R) -> Self {
+        PakFile {
+            pak_version: PakVersion::PakFileVersionInvalid,
+            mount_point: "".to_owned(),
+            compression: CompressionMethod::Unknown,
+            block_size: 0x010000,
+            files: HashMap::new(),
+            reader: Some(BufReader::new(reader)),
+            writer: None,
+        }
+    }
+
+    /// Creates a new `PakFile` configured to write files.
+    pub fn writer(
+        writer: &'data W,
+        pak_version: PakVersion,
+        compression: CompressionMethod,
+    ) -> Self {
+        PakFile {
+            pak_version,
+            mount_point: "../../../".to_owned(),
+            compression,
+            block_size: 0x010000,
+            files: HashMap::new(),
+            reader: None,
+            writer: Some(BufWriter::new(writer)),
+        }
+    }
+
+    /// Load the file info contained in the footer into memory to start reading individual files.
+    pub fn load_files(&mut self) -> Result<(), PakError> {
```

Comment:
Maybe call this `load_index`? Also change the comment to `Load file index info`?